src/frontend/js/common/searchFields/index.tsx

```diff
@@ -0,0 +1,111 @@
+
+/**
+ * Derive from `SearchSuggestion` the kind of suggestion sections our `<SearchSuggestField />` component supports.
+ */
+export type SearchSuggestionSection = SuggestionSection<SearchSuggestion>;
+
+/**
+ * Helper to typecheck `react-autosuggest` expected props. Defines what is acceptable as a suggestion
+ * throughout our code related to this instance of `<Autosuggest />`.
+ */
+export type SearchAutosuggestProps = AutosuggestProps<SearchSuggestion>;
+
+/**
+ * `react-autosuggest` callback to get a human string value from a Suggestion object.
+ * @param suggestion The relevant suggestion object.
+ */
+export const getSuggestionValue: SearchAutosuggestProps['getSuggestionValue'] = suggestion =>
+  suggestion.title;
+
+/**
+ * `react-autosuggest` callback to render one suggestion.
+ * @param suggestion Either a resource suggestion with a model name & a machine name, or the default
+ * suggestion with some text to render.
+ */
+export const renderSuggestion: SearchAutosuggestProps['renderSuggestion'] = suggestion => (
+  <span>{suggestion.title}</span>
+);
+
+/**
+ * `react-autosuggest` callback to build up the list of suggestions and sections whenever user
+ * interaction requires us to create or update that list.
+ * @param filters The general filters object as returned by the API on a course search.
+ * @param setSuggestions The suggestion setter method for the component using our helper.
+ * @param incomingValue The current value of the search suggest form field.
+ */
+export const onSuggestionsFetchRequested = async (
+  filters: APICourseSearchResponse['filters'],
+  setSuggestions: (suggestions: SearchSuggestionSection[]) => void,
+  incomingValue: string,
+) => {
+  if (incomingValue.length < 3) {
+    return setSuggestions([]);
+  }
+
+  // Fetch the suggestions for each resource-based section to build out the sections
+  let sections: SearchSuggestionSection[];
+  try {
+    sections = (await Promise.all(
+      Object.values(await filters)
+        .filter(filterdef => filterdef.is_autocompletable)
+        .map(filterdef =>
+          getSuggestionsSection(
+            filterdef.name,
+            filterdef.human_name,
+            incomingValue,
+          ),
+        ),
+      // We can assert this because of the catch below
+    )) as SearchSuggestionSection[];
+  } catch (error) {
+    return handle(error);
+  }
+
+  setSuggestions(
+    // Drop sections with no results as there's no use displaying them
+    sections.filter(section => !!section!.values.length),
+  );
+};
+
+/**
+ * Helper to pick out the relevant filter for a suggestion from the general filters object.
+ * @param filters The general filters object as returned by the API on a course search.
+ * @param suggestion The suggestion with which we're doing some work (eg. the suggestion the user selected).
+ */
+export const getRelevantFilter = (
+  filters: APICourseSearchResponse['filters'],
+  suggestion: SearchSuggestion,
+) => {
```

Comment:
Some tests are missing for this module, no ?

Reply:
Not sure they're very relevant since everything is tested through the two search field components. There's no new code here, it was just moved around.

Reply:
testing it direcly in a dedicated test is better IMO but if you think it's enough like this we can continue this way.

Reply:
This is just glue code that is very specific to our autocomplete components and IMO it's barely worth it to even split it out to reuse like I did. I think it's better to not test this and test the actual components users will interact with.